File: src/runtime/boxes/atp/bb_atp.c
/* _XATP     ATP         @var — write cursor Δ as DT_I into varname; no backtrack */
#include "bb_box.h"
#include "../../snobol4/snobol4.h"
#include <stdlib.h>
#include <string.h>


spec_t bb_atp(void *zeta, int entry)
{
    atp_t *ζ = zeta;
    spec_t ATP;
    if (entry==α)                                                               goto ATP_α;
    if (entry==β)                                                               goto ATP_β;
    ATP_α:          ζ->done=1;                                                  
                    if (ζ->varname && ζ->varname[0]) {                          
                        DESCR_t v={.v=DT_I,.i=(int64_t)Δ};                      
                        NV_SET_fn(ζ->varname, v); }                             
                    ATP = spec(Σ+Δ,0);                                          goto ATP_γ;
    ATP_β:                                                                      goto ATP_ω;
    ATP_γ:                                                                      return ATP;
    ATP_ω:                                                                      return spec_empty;
}

atp_t *bb_atp_new(const char *varname)
{ atp_t *ζ=calloc(1,sizeof(atp_t)); ζ->varname=varname; return ζ; }

File: src/runtime/boxes/atp/bb_atp.java
package bb;

/**
 * bb_atp.java — ATP: @var — write cursor position Δ as integer into varname
 * Port of bb_atp.c / bb_atp.s
 *
 *   ATP_α:  if (varname) NV_SET(varname, Δ);
 *           ATP=spec(Σ+Δ,0);                    goto ATP_γ;
 *   ATP_β:                                      goto ATP_ω;
 */
public class bb_atp extends bb_box {

    public interface IntSetter { void set(String varname, int value); }

    private final String    varname;
    private final IntSetter setter;

    public bb_atp(MatchState ms, String varname, IntSetter setter) {
        super(ms);
        this.varname = varname;
        this.setter  = setter;
    }

    @Override public Spec α() {
        if (varname != null && !varname.isEmpty())
            setter.set(varname, ms.delta);
        return new Spec(ms.delta, 0);                                          // ATP_γ
    }

    @Override public Spec β() { return null; }                              // ATP_ω
}
